sapi/fpm/fpm/fpm_sockets.c

```diff
@@ -58,40 +59,24 @@ static void fpm_sockets_cleanup(int which, void *arg) /* {{{ */
 	unsigned i;
 	unsigned socket_set_count = 0;
 	unsigned socket_set[FPM_ENV_SOCKET_SET_MAX];
-	unsigned socket_set_buf = 0;
 	char envname[32];
-	char *env_value = 0;
-	int p = 0;
+	smart_string env_str = {0};
 	struct listening_socket_s *ls = sockets_list.data;
 
 	for (i = 0; i < sockets_list.used; i++, ls++) {
 		if (which != FPM_CLEANUP_PARENT_EXEC) {
 			close(ls->sock);
 		} else { /* on PARENT EXEC we want socket fds to be inherited through environment variable */
-			char fd[32];
-			char *tmpenv_value;
-			snprintf(fd, sizeof(fd), "%d", ls->sock);
-
-			socket_set_buf = (i % FPM_ENV_SOCKET_SET_SIZE == 0 && i) ? 1 : 0;
-			tmpenv_value = realloc(env_value, p + (p ? 1 : 0) + strlen(ls->key) + 1 + strlen(fd) + socket_set_buf + 1);
-			if (!tmpenv_value) {
-				zlog(ZLOG_SYSERROR, "failure to inherit data on parent exec for socket `%s` due to memory allocation failure", ls->key);
-				free(ls->key);
-				break;
-			}
-
-			env_value = tmpenv_value;
-
 			if (i % FPM_ENV_SOCKET_SET_SIZE == 0) {
-				socket_set[socket_set_count] = p + socket_set_buf;
+				smart_string_appendc(&env_str, '\0');
```

Comment:
This is master process code - it should not use zend alloc. Ideally all master code should also not immediately exit during the alloc failure so it should keep using standard malloc / realloc / free. It means even the persistent version would not be preferred here. I would prefer not the use smart string at all here